src/script/components/MessagesList/Message/ContentMessage/FailedToSendWarning/FailedToSendWarning.tsx

```diff
@@ -0,0 +1,111 @@
+
+import {useState} from 'react';
+
+import type {QualifiedUserClients} from '@wireapp/api-client/lib/conversation';
+import {QualifiedId} from '@wireapp/api-client/lib/user';
+
+import {Bold, Button, ButtonVariant} from '@wireapp/react-ui-kit';
+
+import {t} from 'Util/LocalizerUtil';
+import {matchQualifiedIds} from 'Util/QualifiedId';
+
+import {warning} from './FailedToSendWarning.styles';
+
+export type User = {id: QualifiedId; name: string};
+type Props = {
+  failedToSend: QualifiedUserClients;
+  knownUsers: User[];
+};
+
+type ParsedUsers = {namedUsers: User[]; unknownUsers: QualifiedId[]};
+
+function generateNamedUsers(users: User[], userClients: QualifiedUserClients): ParsedUsers {
+  return Object.entries(userClients).reduce<ParsedUsers>(
+    (namedUsers, [domain, domainUsers]) => {
+      const domainNamedUsers = Object.keys(domainUsers).reduce<ParsedUsers>(
+        (domainNamedUsers, userId) => {
+          const user = users.find(user => matchQualifiedIds(user.id, {id: userId, domain}));
+          if (user) {
+            domainNamedUsers.namedUsers.push(user);
+          } else {
+            domainNamedUsers.unknownUsers.push({id: userId, domain});
+          }
+          return domainNamedUsers;
+        },
+        {namedUsers: [], unknownUsers: []},
+      );
+      namedUsers.namedUsers.push(...domainNamedUsers.namedUsers);
+      namedUsers.unknownUsers.push(...domainNamedUsers.unknownUsers);
+      return namedUsers;
+    },
+    {namedUsers: [], unknownUsers: []},
+  );
+}
+
+export const FailedToSendWarning = ({failedToSend, knownUsers}: Props) => {
+  const [isOpen, setIsOpen] = useState(false);
+
+  const userCount = Object.entries(failedToSend).reduce(
+    (count, [_domain, users]) => count + Object.keys(users).length,
+    0,
+  );
+
+  const showToggle = userCount > 1;
+
+  const {namedUsers} = generateNamedUsers(knownUsers, failedToSend);
+
+  const message =
+    namedUsers.length === 1
+      ? {head: namedUsers[0].name, rest: t('messageFailedToSendToOne')}
+      : {
+          head: t('messageFailedToSendParticipants', {count: userCount.toString()}),
+          rest: t('messageFailedToSendToSome'),
+        };
+
+  return (
+    <div>
+      <p css={warning}>
+        <Bold css={warning}>{message.head}</Bold> {message.rest}
+      </p>
+      {showToggle && (
+        <>
+          {isOpen && (
+            <p css={warning}>
+              {namedUsers
+                .map(user => (
+                  <span data-uie-name="recipient" data-uie-value={user.id.id} key={user.id.id}>
+                    {user.name}
+                  </span>
+                ))
+                .reduce((prev, element) => {
+                  return prev.length === 0 ? [element] : [...prev, ', ', element];
+                }, [] as any[])}
+              {` ${t('messageFailedToSendWillReceive')}`}
+            </p>
+          )}
+          <Button type="button" variant={ButtonVariant.TERTIARY} onClick={() => setIsOpen(!isOpen)}>
```

Comment:
It's not very safe to toggle the state this way. When we want to set a new state based on the previous one we should use callback instead: `setIsOpen(prev => !prev)`

Reply:
True that, nice one